pkg/controller/master/node/node_remove_controller.go

```diff
@@ -0,0 +1,68 @@
+package node
+
+import (
+	"context"
+
+	"github.com/sirupsen/logrus"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	clusterv1 "sigs.k8s.io/cluster-api/api/v1alpha4"
+
+	"github.com/harvester/harvester/pkg/config"
+	ctlclusterv1 "github.com/harvester/harvester/pkg/generated/controllers/cluster.x-k8s.io/v1alpha4"
+	"github.com/harvester/harvester/pkg/util"
+)
+
+const (
+	nodeRemoveControllerName = "node-remove-controller"
+)
+
+// nodeRemoveHandler force remove machine
+type nodeRemoveHandler struct {
+	machineClient ctlclusterv1.MachineClient
+}
+
+// RemoveRegister registers a controller to remove machine when node is removed
+func RemoveRegister(ctx context.Context, management *config.Management, _ config.Options) error {
+	nodes := management.CoreFactory.Core().V1().Node()
+	machineClient := management.ClusterFactory.Cluster().V1alpha4().Machine()
+	handler := &nodeRemoveHandler{
+		machineClient: machineClient,
+	}
+
+	nodes.OnRemove(ctx, nodeRemoveControllerName, handler.OnNodeRemoved)
+
+	return nil
+}
+
+func (h *nodeRemoveHandler) OnNodeRemoved(_ string, node *corev1.Node) (*corev1.Node, error) {
+	if node == nil || node.Annotations == nil || node.Annotations[clusterv1.MachineAnnotation] == "" {
```

Comment:
we should also check if `DeletionTimestamp == nil` then return immediately.

The logic should only be run when DeletionTimestamp is set.